query/physicalplan/exchange.go

```diff
@@ -0,0 +1,51 @@
+package physicalplan
+
+import (
+	"context"
+
+	"github.com/apache/arrow/go/v8/arrow"
+	"golang.org/x/sync/errgroup"
+)
+
+// ExchangeOperator is used to horizontally scale query execution. It take a
+// single stream of execution and invokes the next step's callback in parallel.
+// The next steps finisher will also be invoked in parallel.
+type ExchangeOperator struct {
+	recordChan chan arrow.Record
+	eg         *errgroup.Group
+	context    context.Context
+}
+
+func Exchange(ctx context.Context) *ExchangeOperator {
+	eg, egCtx := errgroup.WithContext(ctx)
+	return &ExchangeOperator{
+		recordChan: make(chan arrow.Record),
+		eg:         eg,
+		context:    egCtx,
+	}
+}
+
+func (e *ExchangeOperator) Callback(record arrow.Record) error {
+	record.Retain()
+	e.recordChan <- record
+	return nil
+}
+
+func (e *ExchangeOperator) SetNextPlan(nextPlan PhysicalPlan) {
+	// each time a next plan is added invoke a new goroutine to handle the records
+	e.eg.Go(func() error {
```

Comment:
This doesn't seem right. We shouldn't start part of the query execution during planning. We might still end up optimizing the physical plan and reshuffling things, which may further exacerbate the problem.